test/e2e/storage/testsuites/volumeperf.go

```diff
@@ -0,0 +1,295 @@
+var waitForProvisionCh chan []*v1.PersistentVolumeClaim
+
+// InitVolumePerformanceTestSuite returns volumePerformanceTestSuite that implements TestSuite interface
+func InitVolumePerformanceTestSuite() TestSuite {
+	return &volumePerformanceTestSuite{
+		tsInfo: TestSuiteInfo{
+			Name: "volume-lifecycle-performance",
+			TestPatterns: []testpatterns.TestPattern{
+				testpatterns.FsVolModeDynamicPV,
+			},
+		},
+	}
+}
+
+func (t *volumePerformanceTestSuite) GetTestSuiteInfo() TestSuiteInfo {
+	return t.tsInfo
+}
+
+func (t *volumePerformanceTestSuite) SkipRedundantSuite(driver TestDriver, pattern testpatterns.TestPattern) {
+}
+
+func (t *volumePerformanceTestSuite) DefineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	type local struct {
+		config      *PerTestConfig
+		testCleanup func()
+		cs          clientset.Interface
+		ns          *v1.Namespace
+		pvcs        []*v1.PersistentVolumeClaim
+		options     *PerformanceTestOptions
+	}
+	var dInfo *DriverInfo
+	ginkgo.BeforeEach(func() {
+		// Check preconditions
+		dDriver := driver.(DynamicPVTestDriver)
+		if dDriver == nil {
+			framework.Failf("Test driver does not support dynamically created volumes")
+
+		}
+		dInfo = dDriver.GetDriverInfo()
+		if dInfo == nil {
+			e2eskipper.Skipf("Failed to get Driver info -- skipping")
+		}
+		if dInfo.PerformanceTestOptions == nil || dInfo.PerformanceTestOptions.ProvisioningOptions == nil {
+			e2eskipper.Skipf("Driver %s doesn't specify performance test options -- skipping", dInfo.Name)
+		}
+	})
+
+	f := framework.NewDefaultFramework("volume-lifecycle-performance")
+	ginkgo.It("volume lifecycle performance [Slow] [Serial]", func() {
+		l := &local{
+			cs:      f.ClientSet,
+			ns:      f.Namespace,
+			options: dInfo.PerformanceTestOptions,
+		}
+		l.config, l.testCleanup = driver.PrepareTest(f)
+		defer l.testCleanup()
+
+		// Stats for volume provisioning operation
+		// TODO: Add stats for attach, resize and snapshot
+		provisioningStats := &performanceStats{
+			mutex:             &sync.Mutex{},
+			perObjectInterval: make(map[string]*interval),
+			operationMetrics:  &Metrics{},
+		}
+
+		sc := driver.(DynamicPVTestDriver).GetDynamicProvisionStorageClass(l.config, pattern.FsType)
+		ginkgo.By(fmt.Sprintf("Creating Storage Class %s", sc.Name))
+		_, err := l.cs.StorageV1().StorageClasses().Create(context.TODO(), sc, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		defer func() {
+			ginkgo.By(fmt.Sprintf("Deleting Storage Class %s", sc.Name))
+			err := l.cs.StorageV1().StorageClasses().Delete(context.TODO(), sc.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err)
+		}()
+
+		// Create a controller to watch on PVCs
+		// When all PVCs provisioned by this test are in the Bound state, the controller
+		// sends a signal to the channel
+		controller := newPVCWatch(f, l.options.ProvisioningOptions.Count, provisioningStats)
+		stopCh := make(chan struct{})
+		go controller.Run(stopCh)
+		defer close(stopCh)
```

Comment:
Yes, AFAIK closing the channel kills the goroutine

Reply:
This would be worth testing to verify. I encountered issues where `defer`ed cleanup functions did not run when I ctrl+c'd a running test. You can work around this by adding cleanup steps via `f.AddAfterEach()`.

Reply:
@chrishenzie do you have an example usage of `f.AddAfterEach()`? I only see it being used in `NewFramework` in the entire e2e test suite. Alternatively does `f.AfterEach()` solve the problem?

Reply:
See here for an example usage of `AddAfterEach()`:
https://github.com/kubernetes/kubernetes/blob/fb6bc4f8b0eb6805ca9b766722a176a281bd54d0/test/e2e/storage/testsuites/volume_stress.go#L171-L174

`f.AfterEach()` is ran once either a test finishes or is interrupted, and performs cleanup work (like deleting the test namespace). The way to set additional cleanup steps is by using `f.AddAfterEach()`, which will execute these functions when `f.AfterEach()` gets called by ginkgo.

Reply:
done!